src/integration-tests/integration-test.py

```diff
@@ -313,279 +272,315 @@ def check_task(self, task_id: UUID, scalesets: List[Scaleset]) -> Optional[str]:
-            for job_id in self.containers:
-                done: Set[ContainerWrapper] = set()
-                for container in self.containers[job_id]:
-                    if len(container.list_blobs()) > 0:
+                if poll:
+                    print("")
+
+        def check_jobs_impl() -> Tuple[bool, str, bool]:
+            self.cleared = False
+            for job_id in check_containers:
+                finished_containers: Set[Container] = set()
+                for (container_name, container_impl) in check_containers[
+                    job_id
+                ].items():
+                    container_client, count = container_impl
+                    if len(container_client.list_blobs()) >= count:
                         clear()
                         self.logger.info(
-                            "new files in: %s", container.client.container_name
+                            "found files for %s - %s",
+                            jobs[job_id].config.name,
+                            container_name,
                         )
-                        done.add(container)
-                for container in done:
-                    self.containers[job_id].remove(container)
-                if not self.containers[job_id]:
-                    clear()
-                    self.logger.info("finished: %s", self.jobs[job_id].config.name)
-                    finished_job.add(job_id)
+                        finished_containers.add(container_name)
 
-            # check all the tasks associated with the job
-            if self.tasks:
-                scalesets = self.of.scalesets.list()
-                for task_id in self.tasks:
-                    error = self.check_task(task_id, scalesets)
-                    if error is not None:
-                        clear()
-                        self.logger.error(error)
-                        finished_job.add(self.tasks[task_id])
-                        self.failed_jobs.add(self.tasks[task_id])
-
-            # cleanup jobs that are done testing
-            for job_id in finished_job:
-                self.stop_template(
-                    self.jobs[job_id].config.name, delete_containers=False
-                )
+                for container_name in finished_containers:
+                    del check_containers[job_id][container_name]
 
-                for task_id, task_job_id in list(self.tasks.items()):
-                    if job_id == task_job_id:
-                        del self.tasks[task_id]
+            if job_tasks:
+                scalesets = self.of.scalesets.list()
+                for job_id in job_tasks:
+                    finished_tasks: Set[UUID] = set()
+                    for task in job_tasks[job_id]:
+                        if job_id not in jobs:
+                            continue
+                        task_result = self.check_task(jobs[job_id], task, scalesets)
+                        if task_result is None:
```

Comment:
I imagine that some task might never run because of a bug and we will need to know which ones.

I suggest we keep track the list of the pending task here and either log them or return them as part of the result. Also it would be helpful to have an option that fail the check if we have any pending tasks.

Reply:
Addressed